### PR TITLE
[ROCm] fix the performance issue when n=1 or 2

### PR DESCRIPTION
**Pull request by cj401-amd**
## Motivation

on ROCm (or CUDA), slogdet has poor performance when n=1 or 2.  

`jnp.linalg.slogdet(x)` now goes through the analytic path only; gradients use the custom JVP above and never hit the backend algorithms.

### CODE REVIEW

File: jax/_src/numpy/linalg.py [hunk 1]
```diff
@@ -465,6 +465,67 @@ def matrix_rank(
...
 
+def _slogdet_from_det(det: Array, a: Array) -> tuple[Array, Array]:
+  """Helper to compute slogdet from a pre-computed determinant."""
+  dtype = lax.dtype(a)
+  abs_det = ufuncs.abs(det)
+  is_zero = abs_det == jnp.array(0, dtype=dtype)
+  if jnp.iscomplexobj(a):
+    sign = jnp.where(is_zero, jnp.array(0, dtype=dtype), det / ufuncs.astype(abs_det, det.dtype))
+  else:
+    sign = jnp.where(is_zero, jnp.array(0, dtype=dtype), ufuncs.sign(det))
```
Comment (jakevdp) [inline jax/_src/numpy/linalg.py]: I think `ufuncs.sign` has precisely the semantics you're using here – so `sign = ufuncs.sign(a)` should be sufficient regardless of the dtype of `a`.

File: jax/_src/numpy/linalg.py [hunk 2]
```diff
@@ -465,6 +465,67 @@ def matrix_rank(
...
+  logabsdet = jnp.where(
```
Comment (jakevdp) [inline jax/_src/numpy/linalg.py]: Similarly, I think `ufuncs.log` replicates these semantics, so you can do `logabsdet = ufuncs.log(abs_det)`

With these two changes, I don't see much need for this helper function: I'd just compute the `sign(det)` and `log(abs(det))` inline in each of the implementations.

File: jax/_src/numpy/linalg.py (hunk 1 above)
Comment (cj401-amd) [reply jax/_src/numpy/linalg.py]: Thanks for pointing this out @jakevdp.

File: jax/_src/numpy/linalg.py [hunk 3]
```diff
@@ -465,6 +465,65 @@ def matrix_rank(
...
+def _slogdet_1x1(a: Array) -> tuple[Array, Array]:
+  """Analytic slogdet for 1x1 matrices. No LU/solve"""
+  det = a[..., 0, 0]
+  dtype = lax.dtype(a)
+  abs_det = ufuncs.abs(det)
+  is_zero = abs_det == jnp.array(0, dtype=dtype)
+  sign = jnp.where(is_zero, jnp.array(0, dtype=dtype), ufuncs.sign(det))
+  logabsdet = jnp.where(
+      is_zero, jnp.array(-np.inf, dtype=dtype), ufuncs.log(abs_det))
+  return sign, ufuncs.real(logabsdet)
```
Comment (jakevdp) [inline jax/_src/numpy/linalg.py]: I think we can replace lines 473-477 with
```python
return ufuncs.sign(abs_det), ufuncs.real(ufuncs.log(abs_det))
```
All the extra conditions seem consistent with the behavior of `abs` and `log`.

Similar for `_slogdet_2x2`.

File: jax/_src/numpy/linalg.py [hunk 4]
```diff
@@ -465,6 +465,55 @@ def matrix_rank(
...
+def _slogdet_2x2(a: Array) -> tuple[Array, Array]:
+  """Analytic slogdet for 2x2 matrices. No LU/solve."""
+  a00, a01 = a[..., 0, 0], a[..., 0, 1]
+  a10, a11 = a[..., 1, 0], a[..., 1, 1]
+  det = (a00 * a11) - (a01 * a10)
+  abs_det = ufuncs.abs(det)
+  return ufuncs.sign(det), ufuncs.real(ufuncs.log(abs_det))
+
+
+@custom_jvp
```
Comment (jakevdp) [inline jax/_src/numpy/linalg.py:485]: Out of curiosity, why do we need a custom JVP for this? Are there corner cases where standard autodiff returns the wrong result?

Comment (cj401-amd) [reply jax/_src/numpy/linalg.py:485]: Hey @jakevdp, thanks for your comments. 

Custom JVP is required: (1) sign(det) is discontinuous at det=0 so standard autodiff is undefined; we use sign_dot=0 for real and a consistent rule for complex. (2) d/dA log|det(A)| = trace(A⁻¹ Ȧ), which we implement explicitly; autodiff through det→abs→log would not yield this formula and is undefined at singular A.

**Review by jakevdp — COMMENTED** [review]
Looks good – last thing, can you please squash the changes into a single commit? See https://docs.jax.dev/en/latest/contributing.html#single-change-commits-and-pull-requests

**Review by jakevdp — APPROVED** [review]

File: jax/_src/numpy/linalg.py [hunk 5]
```diff
@@ -466,6 +466,55 @@ def matrix_rank(
...
+    raise ValueError(f"_slogdet_small only supports n in {{1, 2}}, got n={n}")
+
+
+def _slogdet_small_jvp(primals, tangents):
+  x, = primals
+  g, = tangents
+  n = x.shape[-1]
+  sign, ans = _slogdet_small(x)
+  if n == 1:
+    ans_dot = ufuncs.real(g[..., 0, 0] / x[..., 0, 0])
```
Comment (hawkinsp) [inline jax/_src/numpy/linalg.py]: We're seeing a test failure in our internal CI on both TPU and B200 GPU for the gradients:
```
NumpyLinalgTest.testSlogdetGrad6 (shape=(1, 1), dtype=<class 'numpy.complex64'>)
...
Not equal to tolerance rtol=0.2, atol=0.1
JVP tangent
Mismatched elements: 1 / 1 (100%)
Max absolute difference among violations: 0.13415217
Max relative difference among violations: 1.
 ACTUAL: array(0.+0.j, dtype=complex64)
 DESIRED: array(0.074208-0.111759j, dtype=complex64)
```

We simply run more test cases by default in the internal CI. I think you might also be able to see this failure if you set `JAX_NUM_GENERATED_CASES=90` (or just force the 1x1 complex64 case by editing the test).

Comment (cj401-amd) [reply jax/_src/numpy/linalg.py]: Hey @hawkinsp, we are still working on to fix the unit test issues.

**Review by jakevdp — APPROVED** [review]
